lib/src/functions_client.dart

```diff
@@ -13,6 +13,26 @@ class FunctionsClient {
   final YAJsonIsolate _isolate;
   final bool _hasCustomIsolate;
 
+  /// Getter for the headers
+  Map<String, String> get headers {
+    return _headers;
+  }
+
+  /// Sets the headers to be sent with requests.
+  /// If an 'Authorization' header is already present,
+  /// it will be kept unless removeAuth is called first.
+  /// @param value a Map containing the new headers to set.
+  set headers(Map<String, String> value) {
+    final authHeader =
+        _headers['Authorization']; // Get the existing Authorization header
+
+    _headers
+      ..clear()
+      ..addAll(Constants.defaultHeaders)
+      ..addAll(authHeader == null ? {} : {'Authorization': authHeader})
```

Comment:
I think we can remove this line to make the behavior a bit more intuitive, but what do you think? If the user wants to remove the auth headers, they can simply omit it from `value`.
```suggestion
```

Reply:
because of ..clear() we need to re-include the auth header after, 
maybe an if else is better than `..addAll(authHeader == null ? {} : {'Authorization': authHeader})`

Reply:
@itisnajim 
99% of the time the user is using this library from supabase-flutter or supabase-dart, which would have the auth header in `value`, so it really doesn't matter, but I feel the inconsistency here. In this case, is the idea to do the same in supabase-dart as well and reserve the auth header? 
https://github.com/supabase/supabase-dart/pull/185#issuecomment-1488580641

Reply:
To me, it just feels a bit counter intuitive to do this to remove something from the header.

```dart
functionsClient.header = {'my-header': 'some-value'};
functionsClient.setAuth(null);
```

It feels natural if auth header was removed with just this.
```dart
functionsClient.header = {'my-header': 'some-value'};
```

Reply:
Ah yes, you are right. I agree with you, so there needs to be the same thing in functionsClient, `storageClient`, and `realtimeClient`

Reply:
@dshukertjr done, i think everything is good, check and let me know https://github.com/supabase/functions-dart/pull/19/files

Reply:
So in order to set your own header, but preserve the existing (including auth), you would need the following?
```dart
functionsClient.headers = functionsClient.headers..['my-header] = 'some-value';
```

Reply:
@Vinzent03 like this:`functionsClient.headers = {...functionsClient.headers, ...{'my-header': 'some-value'}};`

Reply:
I guess some users don't know that the auth cred. are stored in the headers. What do you think about adding a warning to the doc? Something like
```dart
  /// Sets the headers to be sent with requests.
  /// [value] a Map containing the new headers to set.
  ///
  /// Watch out to not accidentally remove the `Authorization` header.
  set headers(Map<String, String> value) {
  ```
What do you think?

Reply:
@Vinzent03 @itisnajim 
Do we actually need the setter here? With just the getter, the users can access all the methods to modify the headers like this:

```dart
// All of these are valid with just exposing the getter
functionsClient.headers.addAll({'my-header': 'my-value'});
functionsClient.headers.clear();
```

I think we might be making things more complicated than needed. We could just have a `headers` getter in all functions, storage, realtime, supabase-dart, and maybe that's enough for developers to do all the headers modification they want to do?

Reply:
The setter was supposed to help us control the headers and ensure the default ones were included. But since users can modify the headers however they want using just the getter, including clearing the map, we need to rethink. We have three options: remove the getter, remove the setter, or create a new method that always includes default headers. The easiest solution might be to just keep the getter.

what do you think guys  😀?

Reply:
I'm casting my vote for just keep the getter. 

Most of the users using `supabase_flutter` won't need to modify the header, and those who needs to modify it will have some what of an advanced use case. Exposing the getter will provide them with the options to do whatever they want with it, with their own responsibilities.

Reply:
@dshukertjr How do you plan on setting the header in sub libraries via supabase-dart/flutter? When we expose a getter on the them as well, it could be tricky, right?

Reply:
Yes it would be like:

```dart
client.headers['some-key'] = 'some-value';
client.realtime.headers['some-key'] = 'some-value';
client.functions.headers['some-key'] = 'some-value';
client.storage.headers['some-key'] = 'some-value';

```

But for the realtime do we have to disconnect and reconnect to apply the headers or there another way for this ...
@dshukertjr @Vinzent03 !?

Reply:
i see it used in `_getAuthHeaders`
https://github.com/supabase/supabase-dart/blob/main/lib/src/supabase_client.dart#L207

Reply:
Yeah, and that's only used for the creation of the sub libraries and the stream method. My new pr https://github.com/supabase/supabase-dart/pull/186 ensures every sub library uses its own headers object, which you can then modify individually.

Reply:
Great, so still have one question of how to apply a new headers to the realtime lib,
because this alone is not enough:
`client.realtime.headers['some-key'] = 'some-value';`

as i said below maybe we have to disconnect and reconnect the socket, or if we can just inject the new headers as like for the access token in
`channel.updateJoinPayload({'user_token': token});`
https://github.com/supabase/realtime-dart/blob/main/lib/src/realtime_client.dart#L336

> Yes it would be like:
> 
> ```dart
> client.headers['some-key'] = 'some-value';
> client.realtime.headers['some-key'] = 'some-value';
> client.functions.headers['some-key'] = 'some-value';
> client.storage.headers['some-key'] = 'some-value';
> ```
> 
> But for the realtime do we have to disconnect and reconnect (i mean the socket connection, like i did earlier: https://github.com/supabase/realtime-dart/pull/63/files ) to apply the headers or there is another way for this ... @dshukertjr @Vinzent03 !?

Reply:
`for a new connect`, so i have to do like ?:
```dart
client.realtime.headers['some-key'] = 'some-value';
if (client.realtime.conn != null){
    // Connect with the new headers.
    client.realtime.disconnect();
    client.realtime.connect();
}
```

Reply:
I don't think that's necessary, because once the connection is established, no http requests are sent and thus no headers. The headers are only used for an initial http request to create the websocket connection. So I see no reason in trying to change the headers for any open connect, because they aren't used, only if you create a new connection.

Reply:
Didn't thought about rls. I don't know realtime good enough for that. Maybe @dshukertjr can help here or can ask the team.

Reply:
I feel like we might be over complicating things a bit here. 

I guess if someone wants to set a custom headers, they can always just initialize Supabase client with the headers option, but @itisnajim what is the primary reason why this does not solve what you want to achieve? The headers set here will be applied to all client libraries, including realtime, so we don't have to worry anything about reconnecting and stuff.

```dart
await Supabase.initialize(
  url: supabaseUrl,
  anonKey: supabaseKey,
  headers: {'device_id': '123'},
);
```

Reply:
I want to be able to change the headers multiple times and at any time, not only in the construction, things should be dynamic

Reply:
Sure. 

As I as said earlier, to update the headers used to access realtime, you can remove channels and re-subscribe again. 

I think if someone wants to dynamically modify the headers , they should handle realtime manually by themselves. 

To complete the implementation that we have, why don't we just expose a getter for `headers` in all of the sub libraries, and introduce a setter and a getter on supabase-dart. Within the method of setter on supabase-dart, we can update each sub libraries' headers.